Alternative savings fix

danjm opened the pull request:
This PR replaces https://github.com/MetaMask/metamask-extension/pull/9662, and fixes some additional issues

The most fundamental change made in this PR is an update to the basic formula used for calculating total savings:

```
average savings = (how much your net worth will change if you use the metaswap best quote) - (how much your net worth would have changed if you had swapped the median quote directly via the aggregator)

where:
(how much your net worth will change if you use the metaswap best quote) = (usd value of tokens received) - (usd network fees) 
and where:

(how much your net worth would have changed if you had swapped the average quote directly via the aggregator) = ((usd value of tokens that would be received from the average quote) / (1 - metamask rate)) - (usd network fees of the average quote)
```
And so the the `savings.total` property is now calculated directly from the overall eth value of the best trade to the overall eth value of the median trade. Related to this is that `savings.performance` is now calculated by subtracting fee savings from total savings.

Another important fix is that what was previously called `ethValueOfTrade` is no longer used in calculations of values that will be shown to the user on the front end. It has been renamed to `tokenValueOfQuoteForSorting` to reflect its purpose. This variable assumes a tokenConversionRate of 1 if none is available, which is fine for comparing quotes to calculate the top quote in such cases, but would be meaningless or misleading if used to calculate user facing values. A new `ethValueOfQuote` variable has been created that defaults the conversion rate to 0 if not available, which will result in ethValueOfQuote being zero (in this case, the UI will have to choose not to render savings data).

Also important change is that `savings.fee` is now calculated by comparing the fee of the best quote to the fee of the quote with the media value, as opposed to comparing the fee of the best quote to the median fee. To support this, the `getMedian` function has been refactored to a function that takes a list of quotes and returns an object with relevant savings information from the quote with the media `ethValueOfQuote`

Finally, this PR ensures that the metamask fee is accounted for in the savings calculations and makes that information available in the savings property for use in the UI

There are some other code refactors made to improve readability given the above changes.

danjm committed:
Alternative savings fix

danjm committed:
Further required changes to savings fix

danjm committed:
Further fix to savings calculations that properly accounts for metamask fees

danjm committed:
metaMaskFeeInEth property on quotes to decimal string

danjm committed:
Fix swaps controller unit tests

danjm committed:
Improve documentation in swaps controller

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -474,30 +469,43 @@ export default class SwapsController {
+      const ethValueOfQuote =
+        destinationToken === ETH_SWAPS_TOKEN_ADDRESS
+          ? decimalAdjustedDestinationAmount.minus(totalEthCost, 10)
+          : new BigNumber(tokenConversionRate || 0, 10)
```
Nit: It has been a little hard to follow all of these `tokenConversionRate || 0` conditions.

Maybe this would be simpler if we split this into three branches, and had all of these calculations in them, e.g.

```
if (destinationToken === ETH_SWAPS_TOKEN_ADDRESS) {
  ...
} else if (tokenConversionRate === undefined) {
  ...
} else {
  ...
}
```

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -507,29 +515,68 @@ export default class SwapsController {
+      // savings.performance is meant to be the eth value of the difference in
+      //   token received when choosing the best quote over the median. Because
+      //   savings.total is a comparison of overall value of the best quote after
+      //   fees have been extracted to overall balue of the media quote before
```
Typo:
```suggestion
      //   fees have been extracted to overall value of the media quote before
```

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -507,29 +515,68 @@ export default class SwapsController {
+      //   metaMaskFeeInEthForBestQuote. This ensures that
+      //   savings.total = savings.performance + savings.fee - savings.metaMaskFee
+      //   Each of these will be represented in the UI
+      savings.performance = savings.total
```
Wouldn't it be simpler to return the `destinationTokenAmount` (or `decimalAdjustedDestinationAmount`) directly from `getMedianEthValueQuote`, instead of having to work backwards to derive it?

Or at least that's my understanding of what's meant here - `bestQuoteDestionationAmount - medianQuoteDestinationAmount - metaMaskFee`

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -507,29 +515,68 @@ export default class SwapsController {
+      //   Each of these will be represented in the UI
+      savings.performance = savings.total
+        .minus(savings.fee, 10)
+        .plus(metaMaskFeeInEthForBestQuote)
```
Shouldn't this be subtracted here, not added? Assuming that we want to compare destination amount _we_ are offering users to the median amount they'd get if they didn't have to pay this fee.

Gudahtt replied:
Oh, also, I'd expect that we'd be subtracting the fee from the median quote here, not adding the fee for the best quote. Maybe I'm missing something here.

danjm replied:
if
`total savings = performance savings + fee savings - metaMaskFee`
then
`performance savings = total savings - fee savings + metaMaskFee`

That is what this calculation of `savings.performance` reflects.

Now, total savings is calculated in this code according to a different formula: `total savings = (eth value of best quote) - (eth value of median quote)`, but by definition I think that both ways of calculating total savings are equivalent.

With all that said, I am currently working on making this code clearer.

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -625,36 +672,44 @@ export default class SwapsController {
       return 0
     }
-    return a.lessThan(b) ? -1 : 1
+    return ethValueOfQuoteA.lessThan(ethValueOfQuoteB) ? -1 : 1
```
Shouldn't this be `isLessThan`? What BigNumber library is being used here?

danjm replied:
In the [changelog for BigNumber version 6.0](https://github.com/MikeMcl/bignumber.js/blob/master/CHANGELOG.md#600), there is `Rename methods: add is prefix to greaterThan, greaterThanOrEqualTo, lessThan and lessThanOrEqualTo`

In our package.json is `"bignumber.js": "^4.1.0"`

Gudahtt replied:
Ah I see! I actually did check the correct version of that library, but the image in the README is updated to reflect the latest version :facepalm:

danjm commented on test/unit/app/controllers/swaps-test.js:
```diff
@@ -763,42 +769,60 @@ describe('SwapsController', function () {
-          median.toNumber(), 3,
-          'should have returned correct median',
+        assert.deepEqual(
+          median, values[1],
```
good question... I'll look into this

danjm committed:
Prevent getMedianEthValueQuote from mutation passed quotes array with .sort() call

danjm committed:
Another fix and refactor to savings calculations in _findTopQuoteAndCalculateSavings

Cleaner structuring of conditionals for setting tokenValueOfQuoteForSorting, ethValueOfQuote and metaMaskFeeInEth in swaps controller

Stop subtracting medianMetaMaskFee from savings, but include it in savings data

Another fix and refactor to savings calculations in _findTopQuoteAndCalculateSavings

danjm committed:
Add and update unit tests for _findTopQuoteAndCalculateSavings

danjm commented:
I just pushed a number of changes, including a rebase onto develop. I will reply to previous comments and add further commentary on latest changes in a little while.

danjm commented:
6dc0b6d216a05cdd6a780699c7a1b41918041253 Added unit tests for _findTopQuoteAndCalculateSavings for the case of ERC-20 -> ERC-20 swaps when the destination conversion rate is known. We should also add unit tests for:

- ETH -> ERC-20 swaps, for both when the destination conversion rate is known and unknown
- ERC-20 -> ERC-20 swaps, for when the destination conversion rate is unknown
- both swaps where ETH and where ERC-20 is being received, and trade.value includes some amount for a fee

So 5 sets of unit tests to add.

danjm commented:
@Gudahtt @rekmarks This is again ready for review.

Some notables changes that have been made:
- I have removed logic that attempts to determine cost "if you had swapped the average quote directly via the aggregator", and have gone back to the old savings calculation based on comparing received tokens and network fees of _our_ best quote to _our_ median quote. Whether we eventually modify this calculation is a UX decision that we are not yet ready to make. For now, for the purposes of analytics on savings, this gives us all the data we need. The metamask fee that would be extracted on the median quote is also now captured, so that we can calculate cost "if you had swapped the average quote directly via the aggregator" from our analytics data if we want. (These changes made in 333a41a93)
- In response to some of @Gudahtt's comments, and in general to try to make the `_findTopQuoteAndCalculateSavings` more understand I have reorganized some of the logic and renamed some variables  (These changes made in 333a41a93)
- The `_findTopQuoteAndCalculateSavings` no longer mutates the quote object it is passed  (These changes made in 333a41a93)
- I have now added many more unit tests for the different scenarios that have to be handled by `_findTopQuoteAndCalculateSavings`  (These changes made in 2217b0235)
- `getMedianEthValueQuote` also no longer mutates the array it is passed  (e8264319e)

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -685,34 +723,51 @@ export default class SwapsController {
+
+  return {
+    ethFee: new BigNumber(quotes[upperIndex].ethFee, 10)
+      .plus(quotes[upperIndex - 1].ethFee)
```
This is missing the radix:

```suggestion
      .plus(quotes[upperIndex - 1].ethFee, 10)
```

Gudahtt replied:
The two `.plus` calls below are missing the radix as well

danjm replied:
Done in 1458ad7d1

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -529,60 +526,101 @@ export default class SwapsController {
+        .toString(10)
       savings.performance = savings.performance.toString(10)
       savings.fee = savings.fee.toString(10)
+      savings.medianMetaMaskFee = medianMetaMaskFee
```
Is this property required for the designs? I'm having trouble imagining where this would be useful.

I don't really understand the purpose of any of these `savings` properties to be honest, apart from `total`. The others seem superfluous.

danjm replied:
For now, `medianMetaMaskFee` is just going to be used in analytics. The[ `performance` and `fee` properties are required for designs](https://github.com/MetaMask/metamask-extension/pull/9666#issuecomment-716700331), and also useful for analytics as we determine the makeup of the total savings for various token pairs and amount (i.e. how much of these savings are due to differences in tokens received vs differences in fee)

danjm committed:
Improve calculation of overallValueOfQuoteForSorting for case where ETH is the source token

danjm commented on app/scripts/controllers/swaps.js:
```diff
@@ -529,60 +526,101 @@ export default class SwapsController {
+      const overallValueOfQuoteForSorting =
+        conversionRateForCalculations === undefined
+          ? ethValueOfTokens
+          : ethValueOfTokens.minus(totalEthCost, 10)
```
Nice catch. Fixed in e52eb94cc

danjm committed:
Clean up getMedianEthValueQuote code, test and comments

danjm commented on app/scripts/controllers/swaps.js:
```diff
@@ -685,34 +723,51 @@ export default class SwapsController {
+  if (quotes.length % 2 === 1) {
     // return middle value
-    return values[(values.length - 1) / 2]
+    return quotes[(quotes.length - 1) / 2]
```
Done in 1458ad7d1

danjm commented on app/scripts/controllers/swaps.js:
```diff
@@ -685,34 +723,51 @@ export default class SwapsController {
- * values. The array will be sorted in place.
- * @returns {import('bignumber.js').BigNumber} The median of the sample.
+ * @param {array} quotes - A sample of quote objects with overallValueOfQuote and ethFee properties
+ * @returns {object} An object with the media overallValueOfQuote for all quotes, and the associated median ethFee and metaMaskFeeInEth.
```
Applied in 1458ad7d1

danjm replied:
I am working on another branch locally that adds typedefs in a number of places, both in this file and in the ui. I will include the use of a typedef in the branch/PR.

danjm commented:
@Gudahtt comments have been addressed

Gudahtt commented on test/unit/app/controllers/swaps-test.js:
```diff
@@ -865,37 +1278,101 @@ describe('SwapsController', function () {
         )
       })
 
       it('calculates median correctly with even sample', function () {
```
This test is... interesting. I was confused at first about why `ethFee` was `20` given that the median was split between 2 and 3, then I realized that one of the `2` quotes was arbitrarily chosen over the other as the source of data.

I think this shows that the median calculation is broken. We should be considering _all_ quotes with the median value as being representative of the median (or all quotes on either side of a split, as in this case), not just one arbitrary selection from the group.

Gudahtt replied:
TIL how statistics work apparently. This was an unexpected discover :sweat_smile:

danjm replied:
oh, wow, very nice catch. The median of the overallValueOfQuote is correctly the average/mean of the middle two numbers, but the other values we calculate are not exactly medians are they... so we should get the mean of all values associated with a overallValueOfQuote that equals either of the middle values used to calculate the median overallValueOfQuote.

This can lead to some interesting edge cases. Imagine 8 quotes, 4 with an overall value of X and 4 with an overall value of Y. The `ethFee`, `metaMaskFeeInEth` and `ethValueOfTokens` values that we associate with the median will just be averages/means of each of those values from each quote.

I think you are right that this is most correct though. And an even number of quote with two identical overall values will be very very rare.

Gudahtt replied:
Actually it's more complicated than that.

```
[
  {
    overallValueOfQuote: '2',
    ethFee: AX,
    metaMaskFeeInEth: AY,
    ethValueOfTokens: AZ,
  },
  {
    overallValueOfQuote: '2',
    ethFee: BX,
    metaMaskFeeInEth: BY,
    ethValueOfTokens: BZ,
  },
  {
    overallValueOfQuote: '3',
    ethFee: CX,
    metaMaskFeeInEth: CY,
    ethValueOfTokens: CZ,
  },
  {
    overallValueOfQuote: '6',
    ethFee: DX,
    metaMaskFeeInEth: DY,
    ethValueOfTokens: DZ,
  },
]
```

In this case, the median is split between `2` and `3`. But we don't want to sum all three `2` quotes along with the `3` quote and divide by 4, because that would too heavily weight the `2` quotes.

Instead I think we should sum all of the `2` quotes and divide by the number of `2` quotes, and do the same for the `3` quotes. And _then_ we'd sum these results and divide by two, as we're doing now.

danjm replied:
Thanks @Gudahtt 

I believe the latest commit correctly handles this, in accordance with both of your comments above. Unit tests are added for these cases as well. ec9f935b9

danjm commented on test/unit/app/controllers/swaps-test.js:
```diff
@@ -350,16 +356,417 @@ describe('SwapsController', function () {
+        assert.deepStrictEqual(resultQuotes, expectedResultQuotes)
+      })
+
+      it('returns the top aggId and quotes with savings and fee values if passed necessary data and the source token is not ETH and an ETH fee is included in the trade value of what would be the best quote', async function () {
```
Some aggregators will require that a fee be paid in ETH, separate from the fee extracted from the destination tokens.

danjm committed:
Clean up _findTopQuoteAndCalculateSavings, create test input and expected results helper functions

danjm commented on test/unit/app/controllers/swaps-test.js:
```diff
@@ -350,16 +356,417 @@ describe('SwapsController', function () {
+      // aggId5: 19.8
+      // aggId6: 19.5
+
+      const baseTestInput = {
```
Done in e4bdc815e

danjm commented on test/unit/app/controllers/swaps-test.js:
```diff
@@ -350,16 +356,417 @@ describe('SwapsController', function () {
     })
 
     describe('_findTopQuoteAndCalculateSavings', function () {
+      let swapsController2
```
Removed in e4bdc815e (and in fact, no such variable needs to be defined in this scope, because the `swapsController` variable defined in the parent scope is set in the respective `beforeEach`)

danjm committed:
Update getMedianEthValueQuote to account for multiple quotes with overall values equal to the median

Gudahtt commented on app/scripts/controllers/swaps.js:
```diff
@@ -685,34 +723,114 @@ export default class SwapsController {
- * @param {import('bignumber.js').BigNumber[]} values - A sample of BigNumber
- * values. The array will be sorted in place.
- * @returns {import('bignumber.js').BigNumber} The median of the sample.
+ * @param {Array} quotes - A sample of quote objects with overallValueOfQuote, ethFee, and metaMaskFeeInEth  properties
+ * @returns {Object} An object with the overallValueOfQuote, ethFee, and metaMaskFeeInEth of the quote with the median overallValueOfQuote
```
```suggestion
 * @param {Array} quotes - A sample of quote objects with overallValueOfQuote, ethFee, ethValueOfTokens, and metaMaskFeeInEth  properties
 * @returns {Object} An object with the ethFee, ethValueOfTokens, and metaMaskFeeInEth of the quote with the median overallValueOfQuote
```

The property `ethValueOfTokens` was missing from both of these, and the `overallValueOfQuote` property was mistakenly included in the description of the return (it is not actually present).

danjm replied:
ethValueOfTokens was added in 9ec3a0d2c and overallValueOfQuote was removed from the return value in 59faf71e1

danjm committed:
Add jsdoc comment for meansOfQuotesFeesAndValue

danjm committed:
Fix jsdoc comment for getMedianEthValueQuote

Gudahtt reviewed (approved):
LGTM!
